### PR TITLE
`Tutorial groups`: Use DTOs across tutorial group endpoints and in integration tests

### PR DESCRIPTION
**Pull request by WoH**
### Summary

Migrate tutorial group REST endpoints from returning JPA entities to dedicated DTO records, update all integration tests to deserialize responses as the new DTO types, and regenerate the OpenAPI spec and TypeScript client.

#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.tum.de/developer/guidelines/language).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.tum.de/developer/development-process#pr-naming-conventions).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.tum.de/developer/guidelines/performance) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.tum.de/developer/guidelines/server-development) and the [REST API guidelines](https://docs.artemis.tum.de/developer/guidelines/rest-api).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The tutorial group REST endpoints returned JPA `@Entity` classes directly, violating the project convention that entities must never appear in API responses. Integration tests deserialized responses as entity classes, which worked by accident due to compatible JSON shapes but hid potential DTO mapping bugs.

### Description

**DTO introduction**
- Add `TutorialGroupResponseDTO`, `TutorialGroupUpdateDataDTO`, and `TutorialGroupsConfigurationDTO` as Java records in the `dto` package.
- Nested records for sub-entities: `CourseDTO`, `UserDTO`, `TutorialGroupScheduleDTO`, `TutorialGroupSessionDTO`, `TutorialGroupFreePeriodDTO`, `TutorialGroupRegistrationDTO`.
- Static `from()` factory methods for entity-to-DTO conversion with lazy-loading safety via `getPersistenceUtil().isLoaded()`.
- Update all four REST resources (`TutorialGroupResource`, `TutorialGroupSessionResource`, `TutorialGroupFreePeriodResource`, `TutorialGroupsConfigurationResource`) to return DTOs.
- Input DTO `TutorialGroupsConfigurationUpdateDTO` provides `toEntity()` for the write path.

**Test migration**
- Update test call sites across integration test files to deserialize HTTP responses as DTO records.
- Replace entity getters (`.getId()`) with record accessors (`.id()`).
- Replace cross-type `.isEqualTo(entity)` assertions with field-by-field comparisons.
- Switch configuration POST/PUT request bodies to `TutorialGroupsConfigurationUpdateDTO`.

**OpenAPI regeneration**
- Regenerate `openapi.yaml` and TypeScript client to reflect the new DTO return types.
~- Also picks up unrelated upstream Hyperion schema changes that were pending regeneration on develop.~

~**Fix broken Hyperion import**~
~- Update `CodeGenerationRequestDTO` → `CodeGenerationRequest` import in the code editor component, broken by the OpenAPI client regeneration.~

E: Rebased

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Tutor
- 1 Student
- 1 Course with tutorial groups and a tutorial groups configuration

1. Log in as **Instructor**, navigate to tutorial group management.
2. Create a new tutorial group, verify it appears in the list with correct details.
3. Update the tutorial group (title, teaching assistant, capacity), verify changes persist.
4. Add a schedule to the tutorial group, verify sessions are generated.
5. Create, update, and delete a tutorial group free period, verify the configuration page reflects changes.
6. Create and update tutorial group sessions, verify session details display correctly.
7. Log in as **Student**, verify the tutorial group listing shows appropriate information (registrations and teaching assistant details hidden).
8. Log in as **Tutor**, verify own tutorial group shows registrations and private information.

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students. 

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2

#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| TutorialGroupResponseDTO.java | 100% (14/14) | ✅ |
| TutorialGroupsConfigurationDTO.java | 90% (9/10) | ✅ |
| TutorialGroupUpdateDataDTO.java | 100% (8/8) | ✅ |
| TutorialGroupResource.java | 95% (245/258) | ✅ |
| TutorialGroupSessionResource.java | 95.5% (126/132) | ✅ |
| TutorialGroupFreePeriodResource.java | 92.5% (86/93) | ✅ |
| TutorialGroupsConfigurationResource.java | 89% (65/73) | ✅ |

## Summary by CodeRabbit

* **New Features**
  * Richer tutorial-group responses including schedules, sessions, registrations, TA details, free-periods, and session DTOs.
  * Session actions (activate, cancel) and session endpoints added to the API client.

* **Improvements**
  * API responses and client models standardized to the new response shape across create/update/get flows.
  * Stronger validation for titles, capacity, online flag; update payloads accept schedule data.

* **Tests**
  * Integration tests updated to assert DTO-based API responses.

### CODE REVIEW

**Comment by WoH** [comment]
@tobias-lippert as far as I can tell, you did the migration to API contracts for tutorial groups. Would you mind taking a look and lmk if this is in line with the [goals](https://docs.artemis.tum.de/developer/guidelines/rest-api)

**Comment by WoH** [comment]
The client code changes picked up changes caused by #11802, #11915 & #12043, maybe it makes sense to split these into a separate PR.

**Comment by tobias-lippert** [comment]
> @tobias-lippert as far as I can tell, you did the migration to API contracts for tutorial groups. Would you mind taking a look and lmk if this is in line with the [goals](https://docs.artemis.tum.de/developer/guidelines/rest-api)

I can have a look at it at the weekend. I'm not really actively involved in Artemis development currently.

File: src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupResponseDTO.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroup;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupFreePeriod;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupRegistration;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupRegistrationType;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupSchedule;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupSession;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupSessionStatus;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record TutorialGroupResponseDTO(@Nullable Long id, @Nullable String title, @Nullable CourseDTO course, @Nullable Integer capacity, @Nullable String campus,
```
Comment (tobias-lippert) [inline src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupResponseDTO.java]: Are they really all nullable? What's a use case when the title shouldn't be set?

File: src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupUpdateDataDTO.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroup;
+import de.tum.cit.aet.artemis.tutorialgroup.domain.TutorialGroupSchedule;
+
+/**
+ * DTO for updating tutorial groups. Builds on the create DTO data with schedule information.
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public record TutorialGroupUpdateDataDTO(@Nullable Long id, String title, TutorialGroupDTO.TeachingAssistantDTO teachingAssistant, @Nullable String additionalInformation,
+        @Nullable Integer capacity, @Nullable Boolean isOnline, @Nullable String language, @Nullable String campus, @Nullable TutorialGroupScheduleDTO tutorialGroupSchedule) {
```
Comment (tobias-lippert) [inline src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupUpdateDataDTO.java]: Again. If there are cases where properties a null, we should rather create specific dto objects and use them there instead of a generic one where everything is nullable

**Review by tobias-lippert — COMMENTED** [review]
In general the migration looks good. I just left two comments

**Review by Claudia-Anthropica — DISMISSED** [review]
@WoH Clean DTO migration. The response/update DTOs are well-structured, the `getPersistenceUtil().isLoaded()` guards for lazy collections are the right call, and the new test for registration number exclusion is a nice touch. Looks good.

**Review by Claudia-Anthropica — DISMISSED** [review]
@WoH Solid DTO migration. The response/update/write DTOs are well-structured, the lazy-loading guards are correct, the validation constraints on the write DTOs look good, and the TA lookup now properly throws on unknown logins. The server-tests CI failure is unrelated (JDBC I/O error in ModelingSubmissionIntegrationTest). Nice work cleaning up the feedback from previous rounds.

File: src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupResponseDTO.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+     *
+     * @param tutorialGroup the tutorial group entity
+     * @return the mapped DTO
+     */
+    public static TutorialGroupResponseDTO from(TutorialGroup tutorialGroup) {
+        return new TutorialGroupResponseDTO(tutorialGroup.getId(), tutorialGroup.getTitle(), CourseDTO.from(tutorialGroup.getCourse()), tutorialGroup.getCapacity(),
+                tutorialGroup.getCampus(), tutorialGroup.getLanguage(), tutorialGroup.getAdditionalInformation(), tutorialGroup.getIsOnline(),
+                UserDTO.from(tutorialGroup.getTeachingAssistant()), tutorialGroupSchedule(tutorialGroup), tutorialGroupSessions(tutorialGroup), registrations(tutorialGroup),
+                tutorialGroup.getChannel(), tutorialGroup.getIsUserRegistered(), tutorialGroup.getIsUserTutor(), tutorialGroup.getNumberOfRegisteredUsers(),
+                tutorialGroup.getTeachingAssistantName(), tutorialGroup.getTeachingAssistantId(), tutorialGroup.getTeachingAssistantImageUrl(), tutorialGroup.getCourseTitle(),
```
Comment (dogusaytok) [inline src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupResponseDTO.java:46]: Even though hidePrivacySensitiveInformation() clears the TA entity, these flattened fields still copy TA metadata into the DTO for students, right? I think we need to clear these transients for non-managers or filter them out here. What do you think?

File: src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java [hunk 4]
```diff
@@ -514,17 +504,113 @@ public ResponseEntity<List<TutorialGroupStudentImportDataDTO>> importRegistratio
...
+        tutorialGroup.setId(tutorialGroupDTO.id());
+        tutorialGroup.setTitle(tutorialGroupDTO.title());
+        tutorialGroup.setAdditionalInformation(tutorialGroupDTO.additionalInformation());
+        tutorialGroup.setCapacity(tutorialGroupDTO.capacity());
+        tutorialGroup.setIsOnline(tutorialGroupDTO.isOnline());
+        tutorialGroup.setLanguage(tutorialGroupDTO.language());
+        tutorialGroup.setCampus(tutorialGroupDTO.campus());
+
+        if (tutorialGroupDTO.teachingAssistant() != null && StringUtils.hasText(tutorialGroupDTO.teachingAssistant().login())) {
+            tutorialGroup.setTeachingAssistant(userRepository.findOneByLogin(tutorialGroupDTO.teachingAssistant().login())
```
Comment (dogusaytok) [inline src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java:529]: If a client accidentally sends the login with spaces (like " tutor1 "), hasText() will pass but the DB lookup might fail. We should .trim() the login string before passing it to findOneByLogin() to prevent false "not found" errors. Maybe a quick test for this edge case would make sense also 🙌

Comment (WoH) [reply src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java:529]: see above

File: src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupResponseDTO.java (hunk 3 above)
Comment (WoH) [reply src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupResponseDTO.java:46]: I think this is already beyond what should be done in a DTO migration conceptually.
Goal is not to reinvent the API here.

File: src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java (hunk 4 above)
Comment (WoH) [reply src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java:529]: I will reopen this one with a minimal DTO conversion

**Comment by dogusaytok** [comment]
I've seen also for line `tutorialGroup.setTutorialGroupSchedule(null)` section in `TutorialGroupResource.java`;

The UI is still sending the schedule data from `assembleCreateOrUpdateTutorialGroupRequest()` (`tutorial-create-or-edit.component.ts`), but this line above nulls it out. This means instructors will think they created a scheduled group, but it will save as an unscheduled one. Maybe this could also be solved @WoH with your new PR?
